web/resource/resolve.go

```diff
@@ -0,0 +1,118 @@
+package resource
+
+import (
+	"context"
+	"io/fs"
+	"net/http"
+
+	"darvaza.org/core"
+	"darvaza.org/x/web"
+)
+
+// Resolve expands on [web.Resolve] using RouteParams() to get the URL params
+// from the HTTP Router additionally to the Route Path.
+// It will fallback to [web.Resolve] if a RouteParams handlers isn't present,
+// and also validates the path.
+func Resolve(req *http.Request) (path string, params RouteParamsTable, err error) {
+	ctx := req.Context()
+
+	if h, ok := RouteParams(ctx); ok {
+		path, params, err = h(ctx)
+	} else if h, ok := web.Resolver(ctx); ok {
+		path, err = h(req)
+	} else {
+		path = req.URL.Path
+	}
+
+	if err == nil {
+		s, ok := web.CleanPath(path)
+		if !ok {
+			err = web.NewStatusBadRequest(core.QuietWrap(fs.ErrInvalid, "%q: invalid path", path))
+		}
+		path = s
+	}
+
+	if err != nil {
+		return "", nil, err
+	}
+
+	return path, params, nil
+}
+
+// RouteParamsTable is a list a URL params
+type RouteParamsTable map[string][]any
+
+// Params lists the names of all stored parameters
+func (params RouteParamsTable) Params() []string {
+	return core.SortedKeys(params)
+}
+
+// Add appends a value for the given parameter
+func (params RouteParamsTable) Add(key string, value any) {
+	params[key] = append(params[key], value)
+}
+
+// Len tells how many parameters are stored in the table
+func (params RouteParamsTable) Len(key string) (int, bool) {
+	if m, ok := params[key]; ok {
+		return len(m), true
+	}
+	return 0, false
+}
+
+// Get returns the first stored value for a parameter
+func (params RouteParamsTable) Get(key string) (any, bool) {
```

Comment:
`First`, `Last` and `All` maybe better than `Get*`?

Reply:
amended accordingly